<a href="https://colab.research.google.com/github/trangnh1314/DA/blob/main/B%E1%BA%A3n_sao_c%E1%BB%A7a_B%E1%BA%A3n_sao_c%E1%BB%A7a_B%E1%BA%A3n_sao_c%E1%BB%A7a_DA_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
pip install isodate

In [3]:
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import json
import warnings
import numpy as np
import statistics as st
import plotly.express as px
import google_auth_oauthlib.flow
import googleapiclient.discovery
import googleapiclient.errors
from IPython.display import JSON
import isodate
from sklearn import metrics
import plotly.graph_objs as go
import plotly.io as pio
import plotly.offline as pyo
from plotly.subplots import make_subplots


warnings.filterwarnings('ignore')

In [4]:
CAvideos=pd.read_csv("/content/drive/Shareddrives/Phân tích dữ liệu R Python/Cuối kỳ/CA_youtube_trending_data.csv")

In [5]:
DEvideos=pd.read_csv("/content/drive/Shareddrives/Phân tích dữ liệu R Python/Cuối kỳ/DE_youtube_trending_data.csv")

In [6]:
FRvideos=pd.read_csv("/content/drive/Shareddrives/Phân tích dữ liệu R Python/Cuối kỳ/FR_youtube_trending_data.csv")

In [7]:
KRvideos=pd.read_csv("/content/drive/Shareddrives/Phân tích dữ liệu R Python/Cuối kỳ/KR_youtube_trending_data.csv")

In [8]:
USvideos=pd.read_csv("/content/drive/Shareddrives/Phân tích dữ liệu R Python/Cuối kỳ/US_youtube_trending_data.csv")

In [9]:
# Combine the DataFrames into one
combined_videos = pd.concat([CAvideos, DEvideos, FRvideos, KRvideos, USvideos])

# Reset the index
combined_videos.reset_index(drop=True, inplace=True)

In [10]:
combined_videos

video_id                                              title  \
0        KX06ksuS6Xo  Diljit Dosanjh: CLASH (Official) Music Video |...   
1        J78aPJ3VyNs  I left youtube for a month and THIS is what ha...   
2        M9Pmf9AB4Mo  Apex Legends | Stories from the Outlands – “Th...   
3        3C66w5Z0ixs                 I ASKED HER TO BE MY GIRLFRIEND...   
4        VIUo6yapDbc  Ultimate DIY Home Movie Theater for The LaBran...   
...              ...                                                ...   
1155802  3OFj6l2tQ9s                       World's Most Dangerous Trap!   
1155803  -BXItmWzWNE                                  skibidi toilet 66   
1155804  2ISoqghadkE  FINAL FANTASY XIV: DAWNTRAIL - Extended Teaser...   
1155805  NevPuZ8pJNQ  Gabito Ballesteros x Natanael Cano - LOU LOU (...   
1155806  Cx5yhKIe8fo  Yaisel LM, Blessd, NLE Choppa - Boy Boy (Remix...   

                  publishedAt                 channelId       channelTitle  \
0        2020-08-11T07:30:02Z  UCZRdNleCgW-BGUJf-bbjzQg     Diljit Dosanjh   
1        2020-08-11T16:34:06Z  UCYzPXprvl5Y-Sf0g4vX-m6g      jacksepticeye   
2        2020-08-11T17:00:10Z  UC0ZV6M2THA81QT9hrVWJG3A       Apex Legends   
3        2020-08-11T19:20:14Z  UCvtRTOMP2TqYqu51xNrqAzg           Brawadis   
4        2020-08-11T15:10:05Z  UCDVPcEbVLQgLZX0Rt6jo34A           Mr. Kate   
...                       ...                       ...                ...   
1155802  2023-10-07T16:00:00Z  UCX6OQ3DkcsbYNE6H8uQQuVA            MrBeast   
1155803  2023-10-19T15:00:07Z  UCsSsgPaZ2GSmO6il8Cb5iGA       DaFuq!?Boom!   
1155804  2023-10-21T10:35:34Z  UCpx2BZg8ABgaDV50sGJtWAg  FINAL FANTASY XIV   
1155805  2023-10-20T01:00:11Z  UCSA0AcT4-0GzJJcZITXb_3Q             LOS CT   
1155806  2023-10-20T16:00:12Z  UCVthEy2Ez1cwThlDQvocytw  Yaisel La Melodia   

         categoryId         trending_date  \
0                10  2020-08-12T00:00:00Z   
1                24  2020-08-12T00:00:00Z   
2                20  2020-08-12T00:00:00Z   
3                22  2020-08-12T00:00:00Z   
4                26  2020-08-12T00:00:00Z   
...             ...                   ...   
1155802          24  2023-10-27T00:00:00Z   
1155803          24  2023-10-27T00:00:00Z   
1155804          20  2023-10-27T00:00:00Z   
1155805          10  2023-10-27T00:00:00Z   
1155806          10  2023-10-27T00:00:00Z   

                                                      tags  view_count  \
0        clash diljit dosanjh|diljit dosanjh|diljit dos...     9140911   
1        jacksepticeye|funny|funny meme|memes|jacksepti...     2038853   
2        Apex Legends|Apex Legends characters|new Apex ...     2381688   
3        brawadis|prank|basketball|skits|ghost|funny vi...     1514614   
4        The LaBrant Family|DIY|Interior Design|Makeove...     1123889   
...                                                    ...         ...   
1155802                                             [None]   138273256   
1155803  skibidi meme|sfm animation|funny animation|gmo...    57321283   
1155804  FF14|FFXIV|Final Fantasy XIV (Video Game)|ファイナ...      380009   
1155805  LOU LOU|natanael cano|gabito ballesteros|lo lo...     5039761   
1155806  Yaisel|Blessd|Boy boy|Boy boy Remix|NLE Choppa...     1705262   

           likes  dislikes  comment_count  \
0         296541      6180          30059   
1         353797      2628          40222   
2         146740      2794          16549   
3         156914      5857          35331   
4          45803       964           2198   
...          ...       ...            ...   
1155802  5198820         0         149259   
1155803  1899661         0         189255   
1155804        0         0              0   
1155805   129633         0           3591   
1155806    97388         0           3439   

                                         thumbnail_link  comments_disabled  \
0        https://i.ytimg.com/vi/KX06ksuS6Xo/default.jpg              False   
1        https://i.ytimg.com/vi/J78aPJ3VyNs/def

In [11]:
combined_videos.shape


(1155807, 16)

Understanding the data

In [12]:
# Checking null values

def null_values(df):
    temp = df.isna().sum()
    temp_1 = round(temp * 100 / df.shape[0], 2)

    return pd.DataFrame((temp, temp_1), index = ['Count', 'Percentage']).T.sort_values('Count', ascending = False)


null_df = null_values(combined_videos).reset_index().rename({'index':'Column_name'}, axis =1)
null_df

Column_name    Count  Percentage
0         description  21269.0        1.84
1            video_id      0.0        0.00
2               title      0.0        0.00
3         publishedAt      0.0        0.00
4           channelId      0.0        0.00
5        channelTitle      0.0        0.00
6          categoryId      0.0        0.00
7       trending_date      0.0        0.00
8                tags      0.0        0.00
9          view_count      0.0        0.00
10              likes      0.0        0.00
11           dislikes      0.0        0.00
12      comment_count      0.0        0.00
13     thumbnail_link      0.0        0.00
14  comments_disabled      0.0        0.00
15   ratings_disabled      0.0        0.00

In [13]:
combined_videos['video_id'].nunique()

155387

In [14]:
# This code shows an example of how a randomly chosen video_id has continuous trending_dates

example_id = combined_videos['video_id'].value_counts().sort_values(ascending = False).index[np.random.randint(10,20)]

combined_videos[combined_videos['video_id'] == example_id]

video_id                   title           publishedAt  \
60349   XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
60548   XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
60756   XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
60960   XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
61820   XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
...             ...                     ...                   ...   
983531  XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
983767  XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
983967  XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
984206  XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   
984407  XA2YEHn-A8Q  TWICE Alcohol-Free M/V  2021-06-09T08:58:12Z   

                       channelId       channelTitle  categoryId  \
60349   UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
60548   UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
60756   UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
60960   UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
61820   UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
...                          ...                ...         ...   
983531  UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
983767  UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
983967  UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
984206  UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   
984407  UCaO6TYtlC8U5ttz62hTrZgg  JYP Entertainment          10   

               trending_date  \
60349   2021-06-10T00:00:00Z   
60548   2021-06-10T00:00:00Z   
60756   2021-06-11T00:00:00Z   
60960   2021-06-11T00:00:00Z   
61820   2021-06-13T00:00:00Z   
...                      ...   
983531  2021-06-14T00:00:00Z   
983767  2021-06-15T00:00:00Z   
983967  2021-06-15T00:00:00Z   
984206  2021-06-16T00:00:00Z   
984407  2021-06-16T00:00:00Z   

                                                     tags  view_count  \
60349   TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    14056784   
60548   TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    23055895   
60756   TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    26519269   
60960   TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    37724565   
61820   TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    62576210   
...                                                   ...         ...   
983531  TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    70540355   
983767  TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    73373057   
983967  TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    76420865   
984206  TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    78550806   
984407  TWICE|트와이스|taste of love|alcoholfree|alcoholfr...    81850570   

          likes  dislikes  comment_count  \
60349   1658150     17444         465217   
60548   1862319     26564         599993   
60756   2195321     32490         650967   
60960   2277582     38225         704607   
61820   2466782     50824         830968   
...         ...       ...            ...   
983531  2528258     54570         870781   
983767  2554867     55856         886556   
983967  2580630     57656         902447   
984206  2603611     58833         917824   
984407  2626749     60378         934021   

                                        thumbnail_link  comments_disabled  \
60349   https://i.ytimg.com/vi/XA2YEHn-A8Q/default.jpg              False   
60548   https://i.ytimg.com/vi/XA2YEHn-A8Q/default.jpg              False   
60756   https://i.ytimg.com/vi/XA2YEHn-A8Q/default.jpg              False   
60960   https://i.ytimg.com/vi/XA2YEHn-A8Q/default.jpg              False   
61820   https://i.ytimg.com/vi/XA2YEHn-A8Q/default.jpg              False   
...                                                ...                ...   
983531  https://i.ytimg.com/vi/XA2YEHn-A8Q/defa

Extracting Duration and Comments from Youtube API

In [15]:

# First let's get unique video ids from the video_id column in dataframe
"""
videos_list = data['video_id'].unique()
# The API_KEY is not displayed for security purposes

api_key = "*********************************"
# This fuction is used to extract 'Duration' and 'Comments' of a provided video_id

def get_video_info(vid_id, api_key):
    api_service_name = "youtube"
    api_version = "v3"

    youtube = googleapiclient.discovery.build(
        api_service_name, api_version, developerKey=api_key)

    request = youtube.videos().list(
        part="snippet,contentDetails,statistics,status",
        id=vid_id
    )

    response = request.execute()

    info_temp = []
    info_temp.append(vid_id)
    info_temp.append(response['items'][0]['contentDetails']['duration'])
    info_temp.append(response['items'][0]['status']['madeForKids'])

    youtube = googleapiclient.discovery.build(
    api_service_name, api_version, developerKey = api_key)

    request = youtube.commentThreads().list(
        part="snippet,replies",
        videoId=vid_id,
        textFormat = "plainText",
        order = "relevance"
    )
    response_comments = request.execute()

    info_temp.append([response_comments['items'][i]['snippet']['topLevelComment']['snippet']['textOriginal'] for i in range(20)])


    return info_temp



# This function is used to create a dataframe from the list of video_ids in video_list and save it in the repository.
# Start and end number is the range of video_ids you want the data for.

def data_to_csv(videos_list, start_number, end_number):
    print("Code Running")
    video_info = []
    deleted_videos = []

    for vid_id in videos_list[start_number : end_number]:
        try:
            video_info.append(get_video_info(vid_id, api_key))
        except:
            deleted_videos.append(vid_id)

    print("Number of videos extracted =", len(video_info))
    print("Number of video deleted =", len(deleted_videos))

    df = pd.DataFrame(video_info, columns = ['video_id', 'Duration', 'madeForKids', 'Comments'])
    df.to_csv('./YoutubeDataFiles/Data_' + str(start_number) + '_' + str(end_number) + '.csv')


# Saving the data to a .csv file by calling the funciton

#data_to_csv(videos_list, 0, 5000)"""


'\nvideos_list = data[\'video_id\'].unique()\n# The API_KEY is not displayed for security purposes\n\napi_key = "*********************************"\n# This fuction is used to extract \'Duration\' and \'Comments\' of a provided video_id\n\ndef get_video_info(vid_id, api_key):\n    api_service_name = "youtube"\n    api_version = "v3"\n\n    youtube = googleapiclient.discovery.build(\n        api_service_name, api_version, developerKey=api_key)\n\n    request = youtube.videos().list(\n        part="snippet,contentDetails,statistics,status",\n        id=vid_id\n    )\n\n    response = request.execute()\n\n    info_temp = []\n    info_temp.append(vid_id)\n    info_temp.append(response[\'items\'][0][\'contentDetails\'][\'duration\'])\n    info_temp.append(response[\'items\'][0][\'status\'][\'madeForKids\'])\n\n    youtube = googleapiclient.discovery.build(\n    api_service_name, api_version, developerKey = api_key)\n\n    request = youtube.commentThreads().list(\n        part="snippet,replie

In [16]:
"""# Let's read the 'extracted files' in and merge them to merge with original Dataframe

extracted_df = pd.concat([pd.read_csv('./YoutubeDataFiles/' + file_csv, index_col = 0) for file_csv in os.listdir('./YoutubeDataFiles')],
                          ignore_index = True)

extracted_df.columns = ['video_id', 'videoDuration', 'madeForKids', 'Comments']

extracted_df.head()"""

"# Let's read the 'extracted files' in and merge them to merge with original Dataframe\n\nextracted_df = pd.concat([pd.read_csv('./YoutubeDataFiles/' + file_csv, index_col = 0) for file_csv in os.listdir('./YoutubeDataFiles')],\n                          ignore_index = True)\n\nextracted_df.columns = ['video_id', 'videoDuration', 'madeForKids', 'Comments']\n\nextracted_df.head()"

Although it was possible to extract all the video_ids at once. It would cost money if you were to extract more than 5000.
Hence, we used different API_KEYS and various start & end_numbers to save the files in the repository.
The files are stored in ./YoutubeDataFiles directory.
The .csv files follow the convention Data_startnumber_endnumber.csv.

In [17]:
# Checking the shape of the extracted dataset

#extracted_df.shape

In [18]:
# Dropping channelId

combined_videos.drop(['channelId'], axis = 1, inplace = True)

# Validating the above code

combined_videos.head(3)

video_id                                              title  \
0  KX06ksuS6Xo  Diljit Dosanjh: CLASH (Official) Music Video |...   
1  J78aPJ3VyNs  I left youtube for a month and THIS is what ha...   
2  M9Pmf9AB4Mo  Apex Legends | Stories from the Outlands – “Th...   

            publishedAt    channelTitle  categoryId         trending_date  \
0  2020-08-11T07:30:02Z  Diljit Dosanjh          10  2020-08-12T00:00:00Z   
1  2020-08-11T16:34:06Z   jacksepticeye          24  2020-08-12T00:00:00Z   
2  2020-08-11T17:00:10Z    Apex Legends          20  2020-08-12T00:00:00Z   

                                                tags  view_count   likes  \
0  clash diljit dosanjh|diljit dosanjh|diljit dos...     9140911  296541   
1  jacksepticeye|funny|funny meme|memes|jacksepti...     2038853  353797   
2  Apex Legends|Apex Legends characters|new Apex ...     2381688  146740   

   dislikes  comment_count                                  thumbnail_link  \
0      6180          30059  https://i.ytimg.com/vi/KX06ksuS6Xo/default.jpg   
1      2628          40222  https://i.ytimg.com/vi/J78aPJ3VyNs/default.jpg   
2      2794          16549  https://i.ytimg.com/vi/M9Pmf9AB4Mo/default.jpg   

   comments_disabled  ratings_disabled  \
0              False             False   
1              False             False   
2              False             False   

                                         description  
0  CLASH official music video performed by DILJIT...  
1  I left youtube for a month and this is what ha...  
2  While running her own modding shop, Ramya Pare...

In [19]:
# Converting 'publishedAt' and 'trending_date' to datetime objects

combined_videos[['publishedAt', 'trending_date']] = combined_videos[['publishedAt', 'trending_date']].apply(lambda x: pd.to_datetime(x))

# Validating the above changes

combined_videos.dtypes[['publishedAt', 'trending_date']]

publishedAt      datetime64[ns, UTC]
trending_date    datetime64[ns, UTC]
dtype: object

In [20]:
# Let's import US_category_id.json file and map the category id's respectively

category_path = '/content/drive/Shareddrives/Phân tích dữ liệu R Python/Cuối kỳ/US_category_id.json'


# Creating a dictionary object which stores the category id and its respective category
category_dict = {}

with open(category_path, 'r') as file:
    json_data = json.load(file)
    for item in json_data['items']:
        category_dict[int(item['id'])] = item['snippet']['title']

combined_videos['categoryId'] = combined_videos['categoryId'].apply(lambda x: category_dict[x])

# Validating the above code
combined_videos['categoryId']

0                   Music
1           Entertainment
2                  Gaming
3          People & Blogs
4           Howto & Style
                ...      
1155802     Entertainment
1155803     Entertainment
1155804            Gaming
1155805             Music
1155806             Music
Name: categoryId, Length: 1155807, dtype: object

In [21]:
# This function converts list to min_max_list as per conditions specified in the above cell

def column_start_end(x: list) -> list:
    return [min(x), max(x)]

def description_col(x: list) -> str:
    temp = pd.Series(list(x))
    temp = temp[[not i for i in temp.isna()]]
    return temp.values[0] if len(temp) else ''

# Grouping by 'video_id' and aggregating using functions as specified in the above cell

modified_df = combined_videos.groupby('video_id').agg({'title':st.mode, 'publishedAt':np.min, 'channelTitle':st.mode, 'categoryId':st.mode,
              'trending_date': column_start_end, 'tags': st.mode, 'likes': column_start_end, 'view_count': column_start_end,
                'comment_count': column_start_end, 'comments_disabled': st.mode, 'ratings_disabled': st.mode,
                                           'description':description_col}).reset_index()

modified_df.head(2)

video_id                               title               publishedAt  \
0  --14w5SOEUs  Migos - Avalanche (Official Video) 2021-06-10 16:00:00+00:00   
1  --2O86Z0hsM            MY TESLA PAYS FOR ITSELF 2022-03-09 23:19:08+00:00   

  channelTitle     categoryId  \
0    MigosVEVO          Music   
1      jf.okay  Entertainment   

                                       trending_date  \
0  [2021-06-11 00:00:00+00:00, 2021-06-17 00:00:0...   
1  [2022-03-11 00:00:00+00:00, 2022-03-16 00:00:0...   

                                                tags             likes  \
0  Migos|Avalanche|Quality|Control|Music/Motown|R...  [122830, 281817]   
1                                             [None]    [16481, 17771]   

           view_count  comment_count  comments_disabled  ratings_disabled  \
0  [1076217, 8034045]  [9539, 17321]              False             False   
1    [466980, 565436]   [1278, 1558]              False             False   

                                         description  
0  Music video by Migos performing Avalanche. Qua...  
1

In [22]:
modified_df.shape

(155387, 13)

The resulting dataframe has 155387 rows and 10 columns.

Let's make the following changes to the resulting dataframe in 8 steps:

Create new columns trending_date_start & trending_date_end from trending_date column.
Create new columns likes_start & likes_end from likes column.
Create new columns view_count_start & view_count_end from view_count column.
tags have [None] values in it. Converting them to Null values.
Create new column tagsCount from tags column. Which indicates the number of tags used in particular video.
Create new column hoursTakenToTrend from trending_date_start & publishedAt columns.
Create new column trendingDaysDuration from trending_date_end & trending_date_start columns.
Drop columns trending_date, likes, publishedAt, view_count as they are no longer required.

In [23]:
# Create new columns trending_date_start & trending_date_end from trending_date column.

modified_df['trending_date_start'] = modified_df['trending_date'].apply(lambda x: min(x))
modified_df['trending_date_end'] = modified_df['trending_date'].apply(lambda x: max(x))

In [24]:
# Create new columns likes_start & likes_end from likes column.

modified_df['likes_start'] = modified_df['likes'].apply(lambda x: min(x))
modified_df['likes_end'] = modified_df['likes'].apply(lambda x: max(x))

In [25]:
# Create new columns `view_count_start` & `view_count_end` from `view_count` column.

modified_df['view_count_start'] = modified_df['view_count'].apply(lambda x: min(x))
modified_df['view_count_end'] = modified_df['view_count'].apply(lambda x: max(x))

In [26]:
# tags have [None] values in it. Converting them to Null values.

modified_df['tags'] = modified_df['tags'].apply(lambda x: x if x!= '[None]' else np.nan)

In [27]:
# Create new column tagsCount from tags column. Which indicates the number of tags used in particular video.

modified_df['tagCount'] = modified_df['tags'].apply(lambda x: 0 if type(x) == float else len(list(x.split('|'))))

In [28]:
modified_df['hoursTakenToTrend'] = round((modified_df['trending_date_start'] - modified_df['publishedAt']).dt.seconds/(60*60), 1)
# Create new column hoursTakenToTrend from trending_date_start & publishedAt columns.


In [29]:
# Create new column trendingDaysDuration from trending_date_end & trending_date_start columns.

modified_df['trendingDaysDuration'] = (modified_df['trending_date_end'] - modified_df['trending_date_start']).dt.days

In [30]:
# Create new columns `comment_count_start` & `comment_count_end` from `comment_count` column.

modified_df['comment_count_start'] = modified_df['comment_count'].apply(lambda x: min(x))
modified_df['comment_count_end'] = modified_df['comment_count'].apply(lambda x: max(x))


In [31]:
# Extracting the day of the week
modified_df['dayofweek'] = modified_df['publishedAt'].dt.day_name()

# Extracting published Year of each video
modified_df['Year'] = modified_df['publishedAt'].dt.year

# Extracting published month of each video
modified_df['Month'] = modified_df['publishedAt'].dt.month

In [32]:
# Drop columns trending_date, likes, publishedAt, view_count as they are no longer required.


In [33]:
modified_df.drop(['trending_date', 'likes', 'publishedAt', 'view_count', 'trending_date_start', 'trending_date_end', 'comment_count'], axis = 1, inplace = True)


In [34]:
# Displaying the dataframe after all the changes

modified_df.head(3)

video_id                                          title channelTitle  \
0  --14w5SOEUs             Migos - Avalanche (Official Video)    MigosVEVO   
1  --2O86Z0hsM                       MY TESLA PAYS FOR ITSELF      jf.okay   
2  --2Us2lFR8Y  JE CONSTRUIS UNE PIÈCE SECRÈTE DANS MA MAISON        ALEKS   

      categoryId                                               tags  \
0          Music  Migos|Avalanche|Quality|Control|Music/Motown|R...   
1  Entertainment                                                NaN   
2  Entertainment  maison|secret|pièce secrète|secrète|constuire|...   

   comments_disabled  ratings_disabled  \
0              False             False   
1              False             False   
2              False             False   

                                         description  likes_start  likes_end  \
0  Music video by Migos performing Avalanche. Qua...       122830     281817   
1                                                           16481      17771   
2  Après avoir réagis à PLEIN de pièces SECRÈTES ...        66822      94116   

   view_count_start  view_count_end  tagCount  hoursTakenToTrend  \
0           1076217         8034045         8                8.0   
1            466980          565436         0                0.7   
2            383690          705147         8                7.0   

   trendingDaysDuration  comment_count_start  comment_count_end  dayofweek  \
0                     6                 9539              17321   Thursday   
1                     5                 1278               1558  Wednesday   
2                     5                 3508               4666   Thursday   

   Year  Month  
0  2021      6  
1  2022      3  
2  2021      2

In [35]:
# Displaying the datatypes of the resultant Dataframe

modified_df.dtypes

video_id                 object
title                    object
channelTitle             object
categoryId               object
tags                     object
comments_disabled          bool
ratings_disabled           bool
description              object
likes_start               int64
likes_end                 int64
view_count_start          int64
view_count_end            int64
tagCount                  int64
hoursTakenToTrend       float64
trendingDaysDuration      int64
comment_count_start       int64
comment_count_end         int64
dayofweek                object
Year                      int64
Month                     int64
dtype: object

In [36]:
modified_df['like/view_start'] = round(modified_df['likes_start'] / modified_df['view_count_start'], 2)

modified_df['like/view_end'] = round(modified_df['likes_end'] / modified_df['view_count_end'], 2)

In [37]:
# Importing necessary libraries

from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.compose import make_column_transformer
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler, StandardScaler
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn import metrics
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.svm import SVR
import wordcloud
import xgboost as xgb
from sklearn.model_selection import GridSearchCV

Preparing data before building the model

In [38]:
# Group the data by video_id and select the maximum values for each column
final_trending_dated_data = combined_videos.groupby('video_id')[['trending_date', 'likes', 'view_count', 'comment_count']].max()

# Merge the maximum values dataframe with the original data based on video_id
modified_df = combined_videos.merge(final_trending_dated_data, on = 'video_id', how = 'left')

# Extract year, month, and weekday from the 'publishedAt' column
modified_df['Year'] = modified_df['publishedAt'].dt.year
modified_df['Month'] = modified_df['publishedAt'].dt.month
modified_df['weekday'] = modified_df['publishedAt'].dt.day_name()

# Calculate the duration between the date when the video starts trending and the date when it stops trending
modified_df['trendingDaysDuration'] = (modified_df['trending_date_y'] - modified_df['trending_date_x']).dt.days

# Remove rows where trendingDaysDuration is 0 and reset the index
modified_df = modified_df[modified_df['trendingDaysDuration'] != 0].reset_index(drop = True)

# Drop columns that are no longer needed
df = modified_df.drop(['video_id', 'title', 'publishedAt', 'tags', 'dislikes',
                    'thumbnail_link', 'description', 'trending_date_x', 'trending_date_y'], axis = 1)

# Rename columns to distinguish start and end values for views, likes, and comments
df = df.rename({'view_count_x': 'view_count_start', 'likes_x': 'likes_start', 'comment_count_x': 'comment_count_start',
          'view_count_y': 'view_count_end', 'likes_y': 'likes_end', 'comment_count_y': 'comment_count_end'}, axis = 1)

Defining Column Transformer for numerical, categorical and boolean columns. We will be using the same preprocessing of data for all the prediction models

In [39]:
def get_preprocessor(X_train, y_train, X_test, y_test):
    # Selecting numerical columns
    numeric_features = X_train.select_dtypes(include = ['int64', 'float64']).columns.values

    # Applying Scaling for numberical columns
    numeric_transformer = Pipeline(
        steps=[("scaler", StandardScaler())]
    )


    # Selecting Categorical columns
    categorical_features = X_train.select_dtypes(include = ['object']).columns.values

    # Applying OneHotEncoder for categorical columns
    categorical_transformer = Pipeline(
        steps=[("encoder", OneHotEncoder(drop = 'first', handle_unknown="ignore"))]
    )


    # Selecting Boolean datatyped columns
    boolean_features = X_train.select_dtypes(include = ['bool']).columns.values
    # Applying binary encoding to those columns
    boolean_transformer = Pipeline(
        steps=[("encoder", OneHotEncoder(drop = 'if_binary', handle_unknown="ignore"))]
    )


    # Building a column transformer which applies all those above mentioned transformations
    preprocessor = ColumnTransformer(
        transformers=[
            ("num", numeric_transformer, numeric_features),
            ("cat", categorical_transformer, categorical_features),
            ("bool", boolean_transformer, boolean_features)
        ]
    )

    return preprocessor

Defining a function that returns metrics for regression models

In [40]:
# Define a function called metric_df that takes a model, training and testing features (X_train, X_test),
# and training and testing target values (y_train, y_test) as inputs
def metric_df_reg(model, X_train, X_test, y_train, y_test):
    # Calculate the R^2 score (coefficient of determination) for the training and testing data
    train_r2 = metrics.r2_score(y_train, model.predict(X_train))
    test_r2 = metrics.r2_score(y_test, model.predict(X_test))

    # Calculate the mean squared error (MSE) for the training and testing data
    train_mse = metrics.mean_squared_error(y_train, model.predict(X_train))
    test_mse = metrics.mean_squared_error(y_test, model.predict(X_test))

    # Calculate the mean absolute error (MAE) for the training and testing data
    train_mae = metrics.mean_absolute_error(y_train, model.predict(X_train))
    test_mae = metrics.mean_absolute_error(y_test, model.predict(X_test))

    # Format the numbers for better readability
    train_r2_str = "{:.4f}".format(train_r2)
    test_r2_str = "{:.4f}".format(test_r2)

    train_mse_str = "{:f}".format(train_mse)
    test_mse_str = "{:f}".format(test_mse)

    train_mae_str = "{:f}".format(train_mae)
    test_mae_str = "{:f}".format(test_mae)

    # Combine the calculated metrics into a 2D NumPy array called 'Metrics'
    Metrics = np.array([[train_r2_str, test_r2_str], [train_mse_str, test_mse_str], [train_mae_str, test_mae_str]])

    # Convert the NumPy array into a Pandas DataFrame with the appropriate column and index labels,
    # and return it
    return pd.DataFrame(Metrics, columns=['Train', 'Test'], index=['R2_score', 'Mean Squared Error', 'Mean Absolute Error'])

Defining a function that returns metrics for categorical models

In [41]:
def metric_df_cat(model, X_train, X_test, y_train, y_test):
    # calculate accuracy scores
    train_accuracy = metrics.accuracy_score(y_train, model.predict(X_train))
    test_accuracy = metrics.accuracy_score(y_test, model.predict(X_test))

    # calculate precision scores
    train_precision = metrics.precision_score(y_train, model.predict(X_train))
    test_precision = metrics.precision_score(y_test, model.predict(X_test))

    # calculate recall scores
    train_recall = metrics.recall_score(y_train, model.predict(X_train))
    test_recall = metrics.recall_score(y_test, model.predict(X_test))

    # create a 2D numpy array with the calculated scores
    Metrics = np.array([[train_accuracy, test_accuracy], [train_precision, test_precision], [train_recall, test_recall]])

    # create and return a pandas DataFrame with the calculated scores
    return pd.DataFrame(Metrics,
                        columns = ['Train', 'Test'], index = ['Accuracy', 'Precision', 'Recall'])

Defining a function that returns a plot that compares all the regression models

In [42]:
# Define Data

def compare_reg_models(models):
    train_r2 = [metrics.r2_score(model.predict(X_train), y_train) for _, model in models.items()]
    test_r2 = [metrics.r2_score(model.predict(X_test), y_test) for _, model in models.items()]

    x_axis = np.arange(len(models))*1.2

    # Multi bar Chart

    fig, ax = plt.subplots(figsize = (10, 3))

    ax.bar(x_axis + (0.20), train_r2, width=0.2, label = 'Train R2 Score')
    ax.bar(x_axis + (0.20 * 2), test_r2, width=0.2, label = 'Test R2 Score')

    for patch in ax.patches:
        ax.annotate(round(patch.get_height(),2), (patch.get_x() + 0.03, patch.get_height() + 0.0032), size = 9)

    # Xticks

    plt.xticks(x_axis + 0.3, models.keys(), rotation = 90)

    # Add legend

    plt.legend(bbox_to_anchor=(0.6, 1.25))

    # Display

    plt.grid()

    plt.show()

Defining a function that returns a plot that compares all the classification models

In [43]:
# Define Data

def compare_clf_models(models):
    train_accuracy = [metrics.accuracy_score(model.predict(X_train), y_train) for _, model in models.items()]
    test_accuracy = [metrics.accuracy_score(model.predict(X_test), y_test) for _, model in models.items()]

    x_axis = np.arange(len(models))*1.2

    # Multi bar Chart

    fig, ax = plt.subplots(figsize = (10, 3))

    ax.bar(x_axis + (0.20), train_accuracy, width=0.2, label = 'Train R2 Score')
    ax.bar(x_axis + (0.20 * 2), test_accuracy, width=0.2, label = 'Test R2 Score')

    for patch in ax.patches:
        ax.annotate(round(patch.get_height(),2), (patch.get_x() + 0.03, patch.get_height() + 0.0032), size = 9)

    # Xticks

    plt.xticks(x_axis + 0.3, models.keys(), rotation = 90)

    # Add legend

    plt.legend(bbox_to_anchor=(0.6, 1.25))

    # Display

    plt.grid()

    plt.show()

In [44]:
mapping = {
    'Monday': 1,
    'Tuesday': 2,
    'Wednesday': 3,
    'Thursday': 4,
    'Friday': 5,
    'Saturday': 6,
    'Sunday': 7,
}
df['weekday'] = df['weekday'].replace(mapping)

In [45]:
# Drop the 'likes_end', 'comment_count_end', and 'trendingDaysDuration' columns from the dataframe
model_df = df.drop(columns = ['likes_end', 'comment_count_end', 'trendingDaysDuration', 'channelTitle'])

# Drop rows containing missing values in place (in the original dataframe)
model_df.dropna(inplace = True)
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Create a LabelEncoder object
le = LabelEncoder()

# Fit the LabelEncoder to the 'categoryId' column
le.fit(model_df['categoryId'])

# Transform the 'categoryId' column using the LabelEncoder
model_df['categoryId_encoded'] = le.transform(model_df['categoryId'])

# Print the encoded and original 'categoryId' columns
# print(df[['categoryId', 'categoryId_encoded']].head())

# Split the data into features (X) and target variable (y)
y = model_df['view_count_end']
X = model_df.drop(columns = ['view_count_end','categoryId'])

In [46]:
y

0          19627350
1           3490530
2           3482753
3           3958226
4           2400821
            ...    
979305      1705262
979306      5039761
979307      1729319
979308       917311
979309    138273256
Name: view_count_end, Length: 979310, dtype: int64

In [47]:
X


view_count_start  likes_start  comment_count_start  comments_disabled  \
0                9140911       296541                30059              False   
1                2038853       353797                40222              False   
2                2381688       146740                16549              False   
3                1514614       156914                35331              False   
4                1123889        45803                 2198              False   
...                  ...          ...                  ...                ...   
979305           1543473        93210                 3409              False   
979306           4130010       122293                 3452              False   
979307           1628081        26137                 4451              False   
979308            913570        14438                 4619              False   
979309         137228041      5185464               149008              False   

        ratings_disabled  Year  Month  weekday  categoryId_encoded  
0                  False  2020      8        2                   7  
1                  False  2020      8        2                   3  
2                  False  2020      8        2                   5  
3                  False  2020      8        2                  10  
4                  False  2020      8        2                   6  
...                  ...   ...    ...      ...                 ...  
979305             False  2023     10        5                   7  
979306             False  2023     10        5                   7  
979307             False  2023     10        5                  13  
979308             False  2023     10        6                  13  
979309             False  2023     10        6                   3  

[979310 rows x 9 columns]

In [48]:
X['comments_disabled'] = X['comments_disabled'].astype(int)
X['ratings_disabled'] = X['ratings_disabled'].astype(int)

In [49]:
# Splitting the data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 50, train_size = 0.8)

# Getting column transformer as a preprocessor
preprocessor = get_preprocessor(X_train, y_train, X_test, y_test)

In [50]:
X

view_count_start  likes_start  comment_count_start  comments_disabled  \
0                9140911       296541                30059                  0   
1                2038853       353797                40222                  0   
2                2381688       146740                16549                  0   
3                1514614       156914                35331                  0   
4                1123889        45803                 2198                  0   
...                  ...          ...                  ...                ...   
979305           1543473        93210                 3409                  0   
979306           4130010       122293                 3452                  0   
979307           1628081        26137                 4451                  0   
979308            913570        14438                 4619                  0   
979309         137228041      5185464               149008                  0   

        ratings_disabled  Year  Month  weekday  categoryId_encoded  
0                      0  2020      8        2                   7  
1                      0  2020      8        2                   3  
2                      0  2020      8        2                   5  
3                      0  2020      8        2                  10  
4                      0  2020      8        2                   6  
...                  ...   ...    ...      ...                 ...  
979305                 0  2023     10        5                   7  
979306                 0  2023     10        5                   7  
979307                 0  2023     10        5                  13  
979308                 0  2023     10        6                  13  
979309                 0  2023     10        6                   3  

[979310 rows x 9 columns]

In [51]:
# preprocessor

Predicting the views by the time video gets out of trending list

SVM

In [ ]:
from sklearn.svm import SVC
# # Let's use Linear regression model and Pipeline the preprocessing and lr model
# svc_model_views = Pipeline(
#     steps=[("preprocessor", preprocessor), ("regressor", SVC())]
# )

# # Training the model
svc_model_views = SVC()
svc_model_views.fit(X_train, y_train)

In [ ]:
metric_df_reg(svc_model_views, X_train, X_test, y_train, y_test)


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import learning_curve

# Define the learning_curves function
def learning_curves(model, X_train, y_train, X_test, y_test):
    train_sizes, train_scores, test_scores = learning_curve(
        model, X_train, y_train, cv=5, scoring='neg_mean_squared_error', train_sizes=np.linspace(0.1, 1.0, 10)
    )

    train_scores_mean = -np.mean(train_scores, axis=1)
    test_scores_mean = -np.mean(test_scores, axis=1)

    plt.figure(figsize=(10, 6))
    plt.plot(train_sizes, train_scores_mean, 'o-', color='r', label='Training error')
    plt.plot(train_sizes, test_scores_mean, 'o-', color='b', label='Testing error')

    plt.title('Learning Curves')
    plt.xlabel('Training examples')
    plt.ylabel('Mean Squared Error')
    plt.legend(loc='best')
    plt.grid(True)
    plt.show()

# Plotting learning curves
learning_curves(svc_model_views, X_train, y_train, X_test, y_test)

Predicting the likes by the time video gets out of trending list


In [ ]:
# # Drop the 'likes_end', 'comment_count_end', and 'trendingDaysDuration' columns from the dataframe
# model_df = df.drop(columns = ['view_count_end', 'comment_count_end', 'trendingDaysDuration'])

# # Drop rows containing missing values in place (in the original dataframe)
# model_df.dropna(inplace = True)

# # Split the data into features (X) and target variable (y)
# y = model_df['likes_end']
# X = model_df.drop(columns = ['likes_end'])

# # Splitting the data into train and test
# X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 50, train_size = 0.8)

# # Getting column transformer as a preprocessor
# preprocessor = get_preprocessor(X_train, y_train, X_test, y_test)

In [ ]:
# Drop the 'likes_end', 'comment_count_end', and 'trendingDaysDuration' columns from the dataframe
model_df = df.drop(columns = ['likes_end', 'comment_count_end', 'trendingDaysDuration', 'channelTitle'])

# Drop rows containing missing values in place (in the original dataframe)
model_df.dropna(inplace = True)
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Create a LabelEncoder object
le = LabelEncoder()

# Fit the LabelEncoder to the 'categoryId' column
le.fit(model_df['categoryId'])

# Transform the 'categoryId' column using the LabelEncoder
model_df['categoryId_encoded'] = le.transform(model_df['categoryId'])

# Print the encoded and original 'categoryId' columns
# print(df[['categoryId', 'categoryId_encoded']].head())

# Split the data into features (X) and target variable (y)
y = model_df['view_count_end']
X = model_df.drop(columns = ['view_count_end','categoryId'])

# Getting column transformer as a preprocessor
preprocessor = get_preprocessor(X_train, y_train, X_test, y_test)

In [ ]:
model_df

SVM

In [ ]:
# # Let's use Linear regression model and Pipeline the preprocessing and lr model
# svm_model_likes = Pipeline(
#     steps=[("preprocessor", preprocessor), ("regressor", LinearRegression())]
# )

svm_model_likes = SVC()
# Training the model
svm_model_likes.fit(X_train, y_train)

In [ ]:
metric_df_reg(svm_model_likes, X_train, X_test, y_train, y_test)

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import learning_curve

# Define the learning_curves function
def learning_curves(model, X_train, y_train, X_test, y_test):
    train_sizes, train_scores, test_scores = learning_curve(
        model, X_train, y_train, cv=5, scoring='neg_mean_squared_error', train_sizes=np.linspace(0.1, 1.0, 10)
    )

    train_scores_mean = -np.mean(train_scores, axis=1)
    test_scores_mean = -np.mean(test_scores, axis=1)

    plt.figure(figsize=(10, 6))
    plt.plot(train_sizes, train_scores_mean, 'o-', color='r', label='Training error')
    plt.plot(train_sizes, test_scores_mean, 'o-', color='b', label='Testing error')

    plt.title('Learning Curves')
    plt.xlabel('Training examples')
    plt.ylabel('Mean Squared Error')
    plt.legend(loc='best')
    plt.grid(True)
    plt.show()
# Training the model
svm_model_likes.fit(X_train, y_train)
# Plotting learning curves
learning_curves(svm_model_likes, X_train, y_train, X_test, y_test)

Predicting the comment_count by the time video gets out of trending list

In [ ]:
# # Drop the 'likes_end', 'comment_count_end', and 'trendingDaysDuration' columns from the dataframe
# model_df = df.drop(columns = ['view_count_end', 'likes_end', 'trendingDaysDuration'])

# # Drop rows containing missing values in place (in the original dataframe)
# model_df.dropna(inplace = True)

# # Split the data into features (X) and target variable (y)
# y = model_df['comment_count_end']
# X = model_df.drop(columns = ['comment_count_end'])

# # Splitting the data into train and test
# X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 50, train_size = 0.8)

# # Getting column transformer as a preprocessor
# preprocessor = get_preprocessor(X_train, y_train, X_test, y_test)

In [ ]:
# Drop the 'likes_end', 'comment_count_end', and 'trendingDaysDuration' columns from the dataframe
model_df = df.drop(columns = ['likes_end', 'comment_count_end', 'trendingDaysDuration', 'channelTitle'])

# Drop rows containing missing values in place (in the original dataframe)
model_df.dropna(inplace = True)

# Create a LabelEncoder object
le = LabelEncoder()

# Fit the LabelEncoder to the 'categoryId' column
le.fit(model_df['categoryId'])

# Transform the 'categoryId' column using the LabelEncoder
model_df['categoryId_encoded'] = le.transform(model_df['categoryId'])

# Print the encoded and original 'categoryId' columns
# print(df[['categoryId', 'categoryId_encoded']].head())

# Split the data into features (X) and target variable (y)
y = model_df['view_count_end']
X = model_df.drop(columns = ['view_count_end','categoryId'])

# Getting column transformer as a preprocessor
preprocessor = get_preprocessor(X_train, y_train, X_test, y_test)

SVM

In [ ]:
# # Let's use Linear regression model and Pipeline the preprocessing and lr model
# svm_model_comments = Pipeline(
#     steps=[("preprocessor", preprocessor), ("regressor", LinearRegression())]
# )
svm_model_comments = SVC()

# Training the model
svm_model_comments.fit(X_train, y_train)

In [ ]:
metric_df_reg(svm_model_comments, X_train, X_test, y_train, y_test)

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import learning_curve

# Define the learning_curves function
def learning_curves(model, X_train, y_train, X_test, y_test):
    train_sizes, train_scores, test_scores = learning_curve(
        model, X_train, y_train, cv=5, scoring='neg_mean_squared_error', train_sizes=np.linspace(0.1, 1.0, 10)
    )

    train_scores_mean = -np.mean(train_scores, axis=1)
    test_scores_mean = -np.mean(test_scores, axis=1)

    plt.figure(figsize=(10, 6))
    plt.plot(train_sizes, train_scores_mean, 'o-', color='r', label='Training error')
    plt.plot(train_sizes, test_scores_mean, 'o-', color='b', label='Testing error')

    plt.title('Learning Curves')
    plt.xlabel('Training examples')
    plt.ylabel('Mean Squared Error')
    plt.legend(loc='best')
    plt.grid(True)
    plt.show()


# Plotting learning curves
learning_curves(svm_model_comments, X_train, y_train, X_test, y_test)

Predicting the number of trending days

In [ ]:
# Drop the 'likes_end', 'comment_count_end', and 'trendingDaysDuration' columns from the dataframe
model_df = df.drop(columns = ['likes_end', 'comment_count_end', 'view_count_end', 'channelTitle'])

# Drop rows containing missing values in place (in the original dataframe)
model_df.dropna(inplace = True)

# Create a LabelEncoder object
le = LabelEncoder()

# Fit the LabelEncoder to the 'categoryId' column
le.fit(model_df['categoryId'])

# Transform the 'categoryId' column using the LabelEncoder
model_df['categoryId_encoded'] = le.transform(model_df['categoryId'])

# Split the data into features (X) and target variable (y)
trendingdays = model_df['trendingDaysDuration']
model_df['trendingDaysDuration'] = model_df['trendingDaysDuration'].apply(lambda x: 0 if (x < 5) else 1)
y = model_df['trendingDaysDuration']
X = model_df.drop(columns = ['trendingDaysDuration', 'categoryId'])



In [ ]:
# Split the data into training and testing sets for model evaluation, with a train size of 0.8 and a random seed of 50
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 50, train_size = 0.8)

# Getting column transformer as a preprocessor
preprocessor = get_preprocessor(X_train, y_train, X_test, y_test)

SVM

In [ ]:
# # Let's use Linear regression model and Pipeline the preprocessing and lr model
# svm_model_trending = Pipeline(
#     steps=[("preprocessor", preprocessor), ("classifier", LogisticRegression())]
# )
svm_model_trending=SVC()
# Training the model
svm_model_trending.fit(X_train, y_train)

In [ ]:
metric_df_cat(svm_model_trending, X_train, X_test, y_train, y_test)

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import learning_curve

# Define the learning_curves function
def learning_curves(model, X_train, y_train, X_test, y_test):
    train_sizes, train_scores, test_scores = learning_curve(
        model, X_train, y_train, cv=5, scoring='neg_mean_squared_error', train_sizes=np.linspace(0.1, 1.0, 10)
    )

    train_scores_mean = -np.mean(train_scores, axis=1)
    test_scores_mean = -np.mean(test_scores, axis=1)

    plt.figure(figsize=(10, 6))
    plt.plot(train_sizes, train_scores_mean, 'o-', color='r', label='Training error')
    plt.plot(train_sizes, test_scores_mean, 'o-', color='b', label='Testing error')

    plt.title('Learning Curves')
    plt.xlabel('Training examples')
    plt.ylabel('Mean Squared Error')
    plt.legend(loc='best')
    plt.grid(True)
    plt.show()


# Plotting learning curves
learning_curves(svm_model_trending, X_train, y_train, X_test, y_test)